semantic/constraint.go

```diff
@@ -0,0 +1,334 @@
+package semantic
+
+import "github.com/influxdata/flux/ast"
+
+// Constraint represents an equality constraint between type expressions
+type Constraint struct {
+	left  TypeVar
+	right Substitutable
+}
+
+// Substitutable represents any type expression containing type variables
+type Substitutable interface {
+	FreeTypeVar() []TypeVar
+}
+
+type arrayTypeExpression struct {
+	elementType Substitutable
+}
+
+func (a arrayTypeExpression) FreeTypeVar() []TypeVar {
+	return a.elementType.FreeTypeVar()
+}
+
+type funcTypeExpression struct {
+	params     map[string]Substitutable
+	returnType Substitutable
+}
+
+func (f funcTypeExpression) FreeTypeVar() []TypeVar {
+	// TODO: Handle duplicates
+	freeTypeVars := make([]TypeVar, 0, 8)
+	for _, v := range f.params {
+		freeTypeVars = append(freeTypeVars, v.FreeTypeVar()...)
+	}
+	return append(freeTypeVars, f.returnType.FreeTypeVar()...)
+}
+
+// ConstraintGenerationVisitor visits a semantic graph and generates
+// constraints between type variables and type expressions.
+type ConstraintGenerationVisitor struct {
+	tenv map[Node]TypeVar
+	cons []Constraint
+}
+
+// Visit a semantic node and generate a type constraint
+func (c *ConstraintGenerationVisitor) Visit(node Node) Visitor {
+	var tv TypeVar
+	var ok bool
+
+	if tv, ok = c.tenv[node]; !ok {
+		return c
+	}
+
+	switch n := node.(type) {
+	case *NativeVariableDeclaration:
+		// Inference Rule: Variable Declaration
+		// ------------------------------------
+		// x = expression
+		//
+		// -> typeof(x) = typeof(expression)
+		//
+		// TODO: variables can change value but not type.
+		// This defines a type constraint.
+		c.cons = append(c.cons, Constraint{
+			left:  c.tenv[n.Identifier],
+			right: c.tenv[n.Init],
+		})
+	case *FunctionExpression:
+		// Inference Rule: Function Expression
+		// -----------------------------------
+		// f = (a, b) => {
+		//     Statement
+		//     Statement
+		//     Return Statement
+		// }
+		//
+		// -> typeof(f) = (typeof(a), typeof(b)) => typeof(Return Statement)
+		funcType := funcTypeExpression{
+			params:     make(map[string]Substitutable, len(n.Params)),
+			returnType: n.returnTypeVar,
+		}
+		c.cons = append(c.cons, Constraint{
```

Comment:
This seems like a special case of Variable Declaration.

Reply:
Sorry, my comment is misleading. There is no variable declaration as part of this rule. I included a variable declaration for readability. It should read `typeof(function expression) = (typeof(a), typeof(b)) => ...`.